python/mlc_chat/gradio.py

```diff
@@ -28,73 +28,179 @@ def _parse_args():
+        stream_interval = gr.update(interactive=True, visible=True)
+        reset_llm_button = gr.update(interactive=True, visible=True)
+        stats_button = gr.update(interactive=True, visible=True)
+        stats_output = gr.update(
+            placeholder="Click to get runtime statistics.", visible=True
+        )
+        text_input = gr.update(interactive=True, placeholder="Type and press enter")
+        llm_buttons = [
+            image_model,
+            stream_interval,
+            reset_llm_button,
+            stats_button,
+            stats_output,
             text_input,
-            gr.update(interactive=True),
-            gr.update(placeholder="Click to get runtime statistics."),
-            gr.update(interactive=True),
-            None,
-            chat_state,
-            img_list,
+        ]
+
+        return llm_buttons + chatbot_vars
+
+    def reload_image_model(self, image_model, chatbot, chat_state, img_list):
+        """Reload the image model."""
+        if image_model == "-None-" or image_model is None:
+            # reload the current llm with original conv template
+            if image_model == "-None-":
+                assert self.curr_llm_lib is not None and self.curr_llm_dir is not None
+                chat_mod.reload(
+                    self.curr_llm_lib, os.path.join(self.curr_llm_dir, "params")
+                )
+                self.reset_runtime_stats()
+                self.process_system_prompts()
+
+            # reset environment and chatbot vars
+            self.vision_exec, self.vision_params, self.vision_embed = None, None, None
+            chatbot = None
+            chat_state = []
+            img_list = []
+            chatbot_vars = [chatbot, chat_state, img_list]
+
+            # change button visibility and state of being interactive
+            image = gr.update(visible=False, interactive=False)
+            text_input = gr.update(interactive=True, placeholder="Type and press enter")
+            image_model_buttons = [image, text_input]
+
+            return image_model_buttons + chatbot_vars
+
+        # get image model dtype
+        quantization_type = image_model.split("-")[-1]
+        if quantization_type[3:5] == "16":
+            self.vision_dtype = "float16"
+        elif quantization_type[3:5] == "32":
+            self.vision_dtype = "float32"
+        else:
+            raise ValueError(f"image model dtype is not supported")
+
+        # reload image model
+        model_dir = os.path.join(self.artifact_path, image_model)
+        self.vision_exec = load_model(model_dir, self.device_name)
+        self.vision_params = load_params(model_dir, self.device)
+        print(f"loaded {image_model}")
+
+        # reload the current llm with MiniGPT's conv template
+        assert self.curr_llm_lib is not None and self.curr_llm_dir is not None
+        app_config_json = '{"conv_template":"minigpt"}'
```

Comment:
Isn't `minigpt` the default `conv_template` in `mlc_chat_config.json`? If so why do we need to override it here?

Reply:
this is because the LLM part in minigpt is actually loaded from vicuna, and when we build vicuna, the conv template is "vicuna_v1.1" by default. In minigpt, the prompts for vicuna is different from the original vicuna, so I choose to override it here.